src/t_hash.c

```diff
@@ -757,64 +779,109 @@ sds hashTypeGetFromHashTable(robj *o, sds field) {
  *
  * If *vll is populated *vstr is set to NULL, so the caller
  * can always check the function return by checking the return value
- * for C_OK and checking if vll (or vstr) is NULL. */
-int hashTypeGetValue(robj *o, sds field, unsigned char **vstr, unsigned int *vlen, long long *vll) {
+ * for C_OK and checking if vll (or vstr) is NULL.
+ *
+ * If field is expired (GET_FIELD_EXPIRED), then it will be lazy deleted.
```

Comment:
Out of curiosity, why are we doing this? Is it needed somewhere?

Reply:
Consider the case of `hincrby` of an expired field. It might increase an obsolete value. Alternatively, the command can implement the logic of checking if the field is expired and make lazy-expiration before using it. In that case, it is better to have a common logic, similar to `lookupKey()` wrt. to keyspace. So this way `hashTypeGetValue()` ended up. Please lmk if you have any comments.

Reply:
Maybe you just need to two versions of `hashTypeGetValue()`. One might expire, other one doesn't. Call it depending on the use case. 

I just have one concern, it increases corner cases a lot. For example, currently, in hmgetCommand():

```c
void hmgetCommand(client *c) {
....

    addReplyArrayLen(c, c->argc-2);
    for (i = 2; i < c->argc; i++) {
        addHashFieldToReply(c, o, c->argv[i]->ptr);
    }
}
```
`addHashFieldToReply()` will call `hashTypeGetValue()`. It might expire the key itself. Then `o` will be dangling pointer in the loop. I see there is a similar problem for `hsetnxCommand()` as well. 

Maybe we should selectively call `hashTypeGetValue()` version that expires key. So, you caller knows the danger.

Reply:
Maybe just to propagate the `GetFieldRes` by function `addHashFieldToReply()` as well.

Reply:
I just seems like most of the time we don't need to expire fields. So, it might be simpler in the code.

Reply:
I think we shouldn't swip it under the rug. There is an added value that everything is one place and it is reflected in the API. Maybe we can have wrapper functions that simplifies it. Similar to the approach of `lookupKey()` family

Reply:
actually i also have the feeling as Ozan, if `o` is implicitly deleted in `hashTypeGetValue`, the caller will need to know the detail implement of `hashTypeGetValue` or need to read the comment for it, otherwise, it's dangerous.